### ISSUE TITLE
Weigths of categorical variables

### ISSUE THREAD

**Issue by erwindassen** [pymc-devs/pymc2#16]
I'm having a problem with Categorical variables and I trying to find out if it a bug of mine or pymc's. It seems that categorical variables are being unevenly weighted. This can be demonstraded in the following program. It models the well-known Monty Hall [example].(http://en.wikipedia.org/wiki/Monty_Hall_problem) See the explanation below.

```
import numpy as np
import pymc as pm

gRange = 3
pRange = 3
mRange = 3

g = pm.Categorical('guest', np.ones(gRange)/gRange, value=2, observed=True)
p = pm.Categorical('prize', np.ones(pRange)/pRange)

# Make a CPT table as an array of CPTColumns
mCPT_shifts = {'guest': 3, 'prize': 1}
mCPT = np.array((  0, 0, 0, 0, 0.5, 1, 0, 1, 0.5,
                 0.5, 0, 1, 0,   0, 0, 1, 0, 0.5,
                 0.5, 1, 0, 1, 0.5, 0, 0, 0,   0), dtype=float)
mCPT.shape = (mRange, pRange*gRange)

parents = {'guest': g, 'prize': p}

# Make a node for variable monty
def m_dist(value=1, CPT_shifts=mCPT_shifts, CPT=mCPT, **parents):
    col = sum(CPT_shifts[par]*parents[par] for par in parents)
    return pm.categorical_like(value, p=CPT[:, col])

m = pm.Stochastic(m_dist, "Monty's door", 'monty', parents, dtype=int,
                  observed=True, value=1)
model = pm.MCMC([g, p, m])
```

The idea is to have two nodes (the prize door and the guest door) as `Categorical` variables with uniform distribution. And the third (Monty's door), depends on the first two. I implement this as a custom `Stochastic` that uses the states (values) of the parents to select a column of the _CPT_ (Conditional Probability Table). It then returns a loglikelyhood for a value according to this table column.

By setting the guest evidence at 2 (door C) and Monty's door at 1 (door B) and tracing the prize door we expect 2/3 ot the prize to be at door A and only 1/3 of the prize to be at door C. Using the snippet:

```
total = 20000
model.sample(iter=total, burn=0, thin=1)
counts = Counter(p.trace[:])
ratio = {var: counts[var]/total for var in counts}
```

I obtain the following results:

`{0: 0.4979, 2: 0.5021}`

Even stranger, if I just select door A as a guest and keep door B for Monty I obtain different ratios (the example should be symmetric here as can be seen from the `mCPT`)!

`{0: 0.20695, 2: 0.79305}`

These numbers are stable under different settings of sample, burning, and thinning. If this is my mistake I welcome corrections. If not, it seems like pymc is drawing from the wrong distribution...

**Comment by erwindassen** [comment]
By the way, if anyone has an idea on how to speed up the samping from the Monty variable please comment. In my real model this is a bottleneck (numba perhaps?).

**Comment by fonnesbeck** [comment]
I'm not clear about a couple of things here: what do you mean by "unevenly weighted"? The probabilities for both Categoricals, as you have set them, is array([ 0.33333333,  0.33333333,  0.33333333]), so they are evenly-weighted in that sense. Moreover, its easy to check whether the distribution is working by checking the log-probabilities for each candidate value:

``` python
In [15]: p.value = 0

In [16]: p.logp
Out[16]: -1.0986122886681098

In [17]: p.value = 1

In [18]: p.logp
Out[18]: -1.0986122886681098

In [19]: p.value = 2

In [20]: p.logp
Out[20]: -1.0986122886681098
```

So, the Categorical appears to be functioning as expected.

One tip: you can use the `@observed` decorator to turn any function that returns a log-probability into an observed stochastic. Thus, your code becomes:

``` python
@pm.observed
def m(value=1, CPT_shifts=mCPT_shifts, CPT=mCPT, **parents):
    col = sum(CPT_shifts[par]*parents[par] for par in parents)
    return pm.categorical_like(value, p=CPT[:, col])
```

However, are you certain this is a valid probability function?

**Comment by erwindassen** [comment]
UPDATE: This was edited to fix the bug reported by fonnesbeck below.

Thanks for the tip. I do think it is a valid probability function (each column adds up to 1). Here is another example. Just two nodes: A,B. B is a categorical depending on A in the sense that the value of A is used to obtain the the probabilities of B via a (deterministic) help function:

```
from pymc import Categorical, deterministic, MCMC
from pylab import hist, show

ITER=5000
BURN = 0
THETA = 0.5

A = Categorical('A', p=[0.5, 0.5])

@deterministic(plot=False)
def B_helper(a=A):
    if A == 1:
        return [THETA, 1 - THETA]
    else:
        return [1 - THETA, THETA]

B = Categorical('B', p = B_helper, value=0, observed=True)
M = MCMC([A,B])

M.sample(iter=ITER, burn=BURN, thin=1)

hist(M.trace('A')[:])
show()

print("\n The probability of A=1 equals: %f" \
    % ((M.trace('A')[:]==1).sum()/float(ITER-BURN)))
```

For `THETA = 0.5` One would expect complete symmetry. Bayes rule gives `P(A=1|B=0) = 0.5`. I obtain something far from it (it appears to converge to 2/3).

Now, if I change the above code by including a "dummy" first sate with probability 0 in the nodes, I do get the expected result:

```
from pymc import Categorical, deterministic, MCMC
from pylab import hist, show

ITER=50000
BURN = 0
THETA = 0.5

A = Categorical('A', p=[0, 0.5, 0.5])  # Note the first state is "dummy" with zero probability.

@deterministic(plot=False)
def B_helper(a=A):
    if A == 0:
        # Shouldn't happen anyway but we don't want to influence the outcome so return uniform
        # on the "non-dummy" states.
        return [0, 0.5, 0.5]
    elif A == 1:
        return [0, THETA, 1 - THETA]  # Note the first state is "dummy" with zero probability.
    elif A == 2:
        return [0, 1 - THETA, THETA]  # Note the first state is "dummy" with zero probability.
    else:
        return [0, 0.5, 0.5]  # If this is not included we get an error!

B = Categorical('B', p = B_helper, value=1, observed=True)
M = MCMC([A,B])

M.sample(iter=ITER, burn=BURN, thin=1)

hist(M.trace('A')[:])
show()

print("\n The probability of A=1 equals: %f" \
    % ((M.trace('A')[:]==1).sum()/float(ITER-BURN)))
```

Here state 0 are dummy states for A and B. And for `THETA = 0.5` I should get (due to symmetry of Bayes rule) `P(A=1|B=1) = 0.5`. I do get that. Furthermore, the position of the "dummy" state is important. If I set the dummy to be state 2 like in the code below I get wrong results again.

```
from pymc import Categorical, deterministic, MCMC
from pylab import hist, show

ITER=50000
BURN = 0
THETA = 0.5

A = Categorical('A', p=[0.5, 0.5, 0])  # Note I changed the order: dummy state is now the third.

@deterministic(plot=False)
def B_helper(a=A):
    if A == 2:
        # Shouldn't happen anyway but we don't want to influence the outcome so return uniform
        # on the "non-dummy" states.
        return [0.5, 0.5, 0]
    elif A == 0:
        return [THETA, 1 - THETA, 0]  # Note I changed the order: dummy state is now the third.
    elif A == 1:
        return [1 - THETA, THETA, 0]  # Note I changed the order: dummy state is now the third.
    else:
        return [0.5, 0.5, 0]  # If this is not included we get an error!

B = Categorical('B', p = B_helper, value=1, observed=True)
M = MCMC([A,B])

M.sample(iter=ITER, burn=BURN, thin=1)

hist(M.trace('A')[:])
show()

print("\n The probability of A=1 equals: %f" \
    % ((M.trace('A')[:]==1).sum()/float(ITER-BURN)))
```

Either I am overseeing something subtle (hopefully not completely obvious) or categoricals are doing something strange in pymc (the logp values seems alright).

**Comment by erwindassen** [comment]
Can somone reproduce this?

**Comment by fonnesbeck** [comment]
I get Pr(A) = 0.5 in the first example, 0 in the second, and 0.333 in the third.

Why does your example include a degenerate element in the categorical?

**Comment by fonnesbeck** [comment]
I get Pr(A) = 0.5 in the first example, 0 in the second, and 0.333 in the third.

One problem I am seeing is that the size of B changes according to the value of A. When A=0, B_helper is a scalar zero, and this is passed to B (which is degenerate); when A=1, or A=2 (the latter is impossible in the last example, since it has zero probability), then B''s argument is a vector of size 3. You can't have the dimension of the model changing at every iteration without implementing a reversible jump algorithm.

**Comment by erwindassen** [comment]
Thanks for the help! I sincerely get Pr(A = 0| B=1) = 1/3 in the first example of the last post.

Of course, you are right about the return value of A. I fixed that but the results stay the same. I've updated the post above with the fixed code so you (or anyone else interested) can reproduce. I get Pr(A=0 | B=1) = 1/3 (incorrect), Pr(A=1 | B=1) = 1/2 (correct), Pr(A=0 | B=1) = 1/3 (incorrect) in the first, second, and third examples respectively. Note that in the second I have to analyse the probability of A = 1 and not A = 0 since this is the "dummy state".

**Comment by fonnesbeck** [comment]
Remember that A is marginal, not conditional. There is nothing in your model that conditions A on B.

**Comment by erwindassen** [comment]
I do set evidence that B=1. I would think that that would make the expected frequencies of A change. Perhaps this intermediate deterministic node is breaking things? On the other hand, that doesn't change the fact that this model (the first one) is symmetric so I can not explain why I consistently get skewed results for A... (1/3).

Is there a recommended way to make such a network of nodes having CPTs? That is, a simple, correct way to define a (parentless) node A with a discrete set of states, each with a probability of ocurring, and such that each of its states determine (deterministically) the probabilities of the possible states of another node (B which is also discrete)? I would then like to set evidence on B and recover information on A. This is exactly what I was trying to achieve with the code in the very first post of this thread (the monty hall example).

What I'm trying to achieve is similar to what is descripteb here: http://stackoverflow.com/questions/22808110/how-to-make-conditional-probability-tables-cpts-for-bayesian-networks-with-pym/22998203#22998203 except that I don't want to "train" the CPT (their values are fixed), perhaps I could make them observed stochastics but that seems a roundabout way of doing it...

Again, thank you for your time. I do appreciate it.

**Comment by fonnesbeck** [comment]
Here is what the DAG looks like for your model:

![dag](http://d.pr/i/1jjxE+)

(its generated by `pymc.graph.dag`, in fact). While it is true that the evidence in B will influence A, the samples of A are marginal, not conditional.

**Comment by erwindassen** [comment]
So there is nothing wrong with PyMC? The value 1/3 for Pr(A=0) (Pr(A=1) = 2/3) in the first of the three examples is correct?

```
[-----------------100%-----------------] 50000 of 50000 complete in 4.0 sec
 The probability of A=1 equals: 0.663500
```

It is difficult for me to see why... if someone can help me understand I appreciate it.

**Comment by fonnesbeck** [comment]
Note that you are extracting the data for `A==0` despite the print statement saying "The probability of A=1 equals:"

**Comment by erwindassen** [comment]
In the first example the code test for A=1 and the result is 2/3 so A=0 has probability 1/3. This is the part I don't understand. The other two examples it is indeed a typo in the str statement. Still, the conclusion that shifting the order of the states should affect the outcome still holds... The ourput of all the examples should be Pr(A == any of the "non-dummy" states) = 1/2. This doesn't happen in the firs examples, happens in the second one and fails in the third one as well...

I will update the examples correting for the typo. Many thanks!

**Comment by erwindassen** [comment]
After a lot of digging I think I found the issue. DiscreteMetropolis accepts a boolean parameter called `positive` which defaults to `True` (essentially `positive=True` takes the absolute value of any negative state as a way of not trowing the sample away). Inspecting the code for it I came to the conclusion that it can lead to a sampling bias. In detail, at each itereation the state of the variable is pertubed by a random walk step. Suppose that the current value os the node is 1 and that the step taken is -2. We end up with the state being equal to -1 which, for default ranges of the `Categorical` (that starts with 0) is invalid. With `positive=True` this is tranformed to 1 which makes this state be more likely than is should. After running the following code (the same as before but setting the step method to be `DiscreteMetropolis` with `positive=False`, I get the right results. Any comment are appreciated.

```
import pymc as pm
from pylab import hist, show

ITER=5000
BURN = 0
THETA = 0.5
B_OBSERVED = True

A = pm.Categorical('A', p=[0.5, 0.5])

@pm.deterministic(plot=False)
def B_helper(a=A):
    if A == 0:
        return [THETA, 1 - THETA]
    elif A == 1:
        return [1 - THETA, THETA]
    else:
        return [0.5, 0.5]

B = pm.Categorical('B', p = B_helper, value=1, observed=B_OBSERVED)
M = pm.MCMC({'A':A, 'B':B})

M.use_step_method(pm.DiscreteMetropolis,
                  M.A, positive=False)

if not B_OBSERVED:
    M.use_step_method(pm.DiscreteMetropolis,
                      M.B, positive=False)

M.sample(iter=ITER, burn=BURN, thin=1)

#hist(M.trace('A')[:])
#show()

print("\n The probability of A=1 equals: %f" \
    % ((M.trace('A')[:]==1).sum()/float(ITER-BURN)))

if not B_OBSERVED:
    print("\n The probability of B=1 equals: %f" \
        % ((M.trace('B')[:]==1).sum()/float(ITER-BURN)))
```

**Comment by kyleabeauchamp** [comment]
@fonnesbeck  I bet this issue was fixed by #85
